### ISSUE TITLE
waitForInvisible produces incorrect/misleading failure messages in certain situations. semantics are unintuitive

### ISSUE THREAD

**Issue by JLRishe** [Codeception/CodeceptJS#943]
I recently discovered that a number of my company's tests were failing on account of an incomplete understanding of how `waitForInvisible` works.

My tester writing the tests quite reasonably assumed that `waitForInvisible` would wait until the page contained _no visible elements_ matching the indicated selector.

After reading through the source code, I now realize that what it actually does is wait until there is _at least one non-visible element_ matching the indicated selector.

This means that if the page contains no elements matching the selector (e.g. if they had been removed from the page or never been there in the first place), `waitForInvisible` eventually indicates a failure with a factually incorrect message:

```js
I.waitForInvisible('#idonotexist');
```

The line above will wait and eventually fail with the message: `element (#idonotexist) still visible after 5 sec`. This is factually incorrect and misleading to someone reading the test results because `#idonotexist` is not "still visible". It doesn't exist at all!

Conversely, if the page contains a mixture of visible and non-visible elements matching a particular selector, `waitForInvisible` will succeed as long as at least one of them is non-visible.

```html
<div class="myclass" style="display: none">I am not visible</div>
<div class="myclass">I am visible</div>
```

```js
I.waitForInvisible('.myclass');   // succeeds
````

I realize that there is a `waitForStalenessOf` method which checks for the non-presence of elements in the DOM (at least this is true in the WebDriverIO helper; in the Protractor helper, it's just an alias for `waitForInvisible`). However, I would think that from a testing perspective, it would be more common and generally better practice to simply test whether an element is visible in the page or not, and not get mired down in the implementation details of hidden vs. not present in the DOM. But Codecept doesn't seem to provide any method for simply checking whether the page does or does not have a particular visible element.

So my questions are:

- Is `.waitForInvisible`'s behavior, as described and demonstrated above, really intentional? It seems strange and counterintuitive to me.
- If it _is_ intentional, can the failure message at least be improved to more clearly convey what it's checking for? E.g. `no invisible element (#idonotexist) found after 5 sec`

### Details

* CodeceptJS version: 1.1.4
* NodeJS Version:  8.9.4
* Operating System: Win10
* WebDriverIO version: 4.10.2
* Configuration file:  unchanged from default

**Comment by reubenmiller** [comment]
You raise some valid points. I can investigate this and maybe look at adding some options which can wait for just 1 matching eememts or all of them. 

I know the next release should fix some of the inconsistencies with the waitUntiExists and waitForStaleness across the different helpers.

I’ll investigate it more and get back to you. Though PRs are also welcome :)

**Comment by JLRishe** [comment]
@reubenmiller Thanks for replying. I'm happy to make a PR, but wanted to confirm whether or not the current behavior is indeed by design before trying to make any changes. If it were up to me, I would have `waitForInvisible` succeed if the page contained no elements matching the selector, but as it's written right now, it [deliberately continues to wait](https://github.com/Codeception/CodeceptJS/blob/99661ff4e4d14390703acf79a609eb9d035f5900/lib/helper/WebDriverIO.js#L1579) if no elements are found:

```js
      const res = await this.browser.elements(withStrictLocator.call(this, locator));
      if (!res.value || res.value.length === 0) return false;
```

How `waitForInvisible` handles the situation where there are multiple matching elements is far less of an issue for me, but I decided to mention that above in the interest of completeness.

**Comment by reubenmiller** [comment]
Personally I think it would be wise to have an option where the user can specify whether the assertion should fail if no matches are found or not. I think the default should fail if no matches are found, so that it ensures the user has the correct selector and the error is not silently passing.

**Comment by JLRishe** [comment]
I see your reasoning there. 

So are you proposing that there be some global configuration setting for that? If it were added as an additional parameter to the method, it would force the test writer to explicitly specify the timeout value every time they wanted to use that option, so having it as a new method parameter probably isn't such a great way to go.

**Comment by reubenmiller** [comment]
So the requirements of the feature would be something like:

* Default setting in the helper config (should be implemented in each helper)
* Function should take an additional option (overriding the default setting). If no option is specified then it takes the configured default

**Comment by JLRishe** [comment]
@reubenmiller Thanks again for clarifying. So just to make sure I understand what you have in mind for the second bullet, the function's new signature would be:

```js
async waitForInvisible(locator, sec = null, newOptionWithNameTBD = null) {
```
and if someone wanted to use this new option when calling the function but still use the default timeout duration, they would call:

```js
I.waitForInvisible('#myselector', null, true)
```

Is that correct? Do you have any thoughts about what to call the new option? Anything I've been able to think of has been pretty wordy.

**Comment by reubenmiller** [comment]
Yeah that looks ok to me. I don't think a verbose name is a bad thing, maybe something like `failIfNotPresent` or `failOnEmptyMatch`...

Though when defaulting to default setting in the helper config, make sure you don't use any references to `this` in the 

Don't use this references in the function declaration

```js
async waitForInvisible(locator, sec = null, newOptionWithNameTBD = this.config.failOnEmptyMatch) {
 ...
}
```

Instead use assign defaults within the function body.

```js
async waitForInvisible(locator, sec = null, newOptionWithNameTBD = null) {
   if (newOptionWithNameTDB == null) {
      newOptionWithNameTDB = this.config.Wai
   }
   ...
}
```
The reason being that references to `this` will break the `codecept list` function.

**Comment by JLRishe** [comment]
@reubenmiller Great. Thanks for the clarifications. I will try to work on this this weekend.

**Comment by reubenmiller** [comment]
Fixed in v1.1.6
